### PR TITLE
WIP Add Hashable derive macro for trait

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This provides a macro for the hashable trait

## Motivation and Context
So we dont have to manually impl hashable

## How Has This Been Tested?
unit tests

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/blockchain/src/blockchainstate.rs [hunk 1]
```diff
@@ -22,11 +22,20 @@
...
 /// The BlockchainState struct keeps record of the current UTXO, total kernels and headers.
-pub struct BlockchainState {}
+use merklemountainrange::mmr::*;
+use tari_core::transaction::{TransactionKernel, TransactionOutput};
+
+type BlockHash = Blake2b;
+
+pub struct BlockchainState {
+    _outputs: MerkleMountainRange<TransactionOutput, BlockHash>,
+    _kernals: MerkleMountainRange<TransactionKernel, BlockHash>,
```
Comment (sdbondi) [inline base_layer/blockchain/src/blockchainstate.rs]:
```suggestion
    _kernels: MerkleMountainRange<TransactionKernel, BlockHash>,
```

File: base_layer/blockchain/src/blockchainstate.rs [hunk 2]
```diff
@@ -22,11 +22,20 @@
...
+}
 
 impl BlockchainState {
     pub fn new() -> BlockchainState {
-        BlockchainState {}
+        BlockchainState { _outputs: MerkleMountainRange::new(), _kernals: MerkleMountainRange::new() }
```
Comment (sdbondi) [inline base_layer/blockchain/src/blockchainstate.rs]:
```suggestion
        BlockchainState { _outputs: MerkleMountainRange::new(), _kernels: MerkleMountainRange::new() }
```

File: infrastructure/derive/src/lib.rs [hunk 3]
```diff
@@ -33,3 +35,108 @@ pub fn derive_hashable_ordering(tokens: TokenStream) -> TokenStream {
     };
     gen.into()
 }
+
+/// This macro will provide a hasable implementation to the a given struct
+#[proc_macro_derive(Hashable, attributes(Digest, hash))]
```
Comment (sdbondi) [inline infrastructure/derive/src/lib.rs]: Some documentation and a quick code example of how to use the derive macro would help the reader here - over and above the unit tests

File: infrastructure/derive/src/lib.rs [hunk 4]
```diff
@@ -33,3 +35,108 @@ pub fn derive_hashable_ordering(tokens: TokenStream) -> TokenStream {
...
+            match item.fields {
+                Fields::Named(ref fields) => {
+                    let recurse = fields.named.iter().map(|f| {
+                        let mut do_we_ignore_field = false;
+                        for attr in &f.attrs {
+                            match attr.interpret_meta().unwrap() {
+                                syn::Meta::NameValue(ref val) => {
+                                    if val.ident.to_string() == "hash" {
+                                        if let syn::Lit::Str(lit) = &val.lit {
+                                            if lit.value() == "Hashable_ignore" {
```
Comment (sdbondi) [inline infrastructure/derive/src/lib.rs]: If I understand this correctly - the usage is

```rust
#[derive(Hashable)]
struct MyThing { 
  #[hash(Hashable_ignore)]
  foo: String,
}
```

Perhaps a better syntax would be

```rust
#[derive(Hashable)]
struct MyThing {
  #[hash(ignore)] /* or */ #[hashable(ignore)]
  foo: String,
}
```

File: infrastructure/merklemountainrange/tests/unit/mmr.rs [hunk 5]
```diff
@@ -24,17 +24,18 @@ use crate::support::{hashvalues::HashValues, testobject::TestObject};
...
-        let object: TestObject<Blake2b> = TestObject::new(i.to_string());
+        let object: TestObject = TestObject::new(i.to_string());
         mmr.add_single(object);
     }
     mmr
 }
 
 #[test]
 fn create_small_mmr() {
+    dbg!(1);
```
Comment (sdbondi) [inline infrastructure/merklemountainrange/tests/unit/mmr.rs]: Clean up

File: infrastructure/derive/src/lib.rs [hunk 6]
```diff
@@ -1,11 +1,13 @@
 #![recursion_limit = "128"]
+//#![feature(Hashable_ignore)]
```
Comment (sdbondi) [inline infrastructure/derive/src/lib.rs]: Clean up

File: infrastructure/merklemountainrange/tests/support/testobject.rs [hunk 7]
```diff
@@ -20,25 +20,28 @@
...
-    fn get_hash(&self) -> ObjectHash {
-        let mut hash = D::new();
-        hash.input(self.id.as_bytes());
-        hash.result().to_vec()
+impl TestObject {
+    pub fn new(id: String) -> TestObject {
+        TestObject { id, string_we_should_not_hash: "a".to_owned() }
     }
 }
+// impl Hashable for TestObject {
```
Comment (sdbondi) [inline infrastructure/merklemountainrange/tests/support/testobject.rs]: Clean up

File: infrastructure/derive/src/lib.rs (hunk 4 above)
Comment (SWvheerden) [reply infrastructure/derive/src/lib.rs]: agreed, went with :     `#[Hashable(Ignore)]`

**Comment by SWvheerden** [comment]
#167

File: infrastructure/derive/src/lib.rs [hunk 8]
```diff
@@ -33,3 +37,112 @@ pub fn derive_hashable_ordering(tokens: TokenStream) -> TokenStream {
...
+// this function processes the individual fields of the hashable trait macro: derive_hashable
+fn handle_fields_for_hashable(item: &Data) -> proc_macro2::TokenStream {
+    match item {
+        Data::Struct(ref item) => {
+            match item.fields {
+                Fields::Named(ref fields) => {
+                    let recurse = fields.named.iter().map(|f| {
+                        let mut do_we_ignore_field = false;
+                        for attr in &f.attrs {
+                            match attr.interpret_meta().unwrap() {
```
Comment (neonknight64) [inline infrastructure/derive/src/lib.rs]: Should this unwrap not be handled differently?

Comment (SWvheerden) [reply infrastructure/derive/src/lib.rs]: This should not break, but in the odd case it does, its only going to break on compile as this is when this code runs, this wont break at runtime

**Comment by SWvheerden** [comment]
Closing this in favour of #212

**Pull request closed without merging**